core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```diff
@@ -0,0 +1,129 @@
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.util.regex.Pattern;
+import org.apache.log4j.BasicConfigurator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JavaGrepImp implements JavaGrep {
+  final Logger logger = LoggerFactory.getLogger(JavaGrep.class);
+
+  private String regex;
+  private String rootPath;
+  private String outFile;
+
+  @Override
+  public void process() throws IOException {
+    List<String> matchedLines = new ArrayList<>();
+    for (File file:listFiles(rootPath)) {
+      for (String line: readLines(file)) {
+        if (containsPattern(line)) {
+          matchedLines.add(line);
+        }
+      }
+    }
+    writeToFile(matchedLines);
+  }
+
+  @Override
+  public List<File> listFiles(String rootDir) {
+    List<File> files = new ArrayList<>();
+    File dir = new File(rootDir);
+    File[] curFiles = dir.listFiles();
+    for (File f: curFiles) {
+      if (f.isFile()) {
+        files.add(f);
+      }
+    }
+    return files;
+  }
+
+  @Override
+  public List<String> readLines(File inputFile) throws IOException{
+    String line;
+    List<String> lines = new ArrayList<>();
+    try (BufferedReader br = new BufferedReader(new FileReader(inputFile))) {
+      line = br.readLine();
+      while(line != null) {
+        lines.add(line);
+        line = br.readLine();
+      }
+    } catch (IOException ex) {
+      logger.error("Cannot read lines from the given file", ex);
+    }
+    return lines;
+  }
+
+  @Override
+  public boolean containsPattern(String line) {
+    return Pattern.matches(getRegex(), line);
+  }
+
+  @Override
+  public void writeToFile(List<String> lines) throws IOException {
+    BufferedWriter wr = new BufferedWriter(new FileWriter(getOutFile()));
+    for (String line: lines) {
+      wr.write(line + "\n");
+    }
+    wr.close();
+  }
+
+  @Override
+  public String getRootPath() {
+    return this.rootPath;
+  }
+
+  @Override
+  public void setRootPath(String rootPath) {
+    this.rootPath = rootPath;
+  }
+
+  @Override
+  public String getRegex() {
+    return this.regex;
+  }
+
+  @Override
+  public void setRegex(String regex) {
+    this.regex = regex;
+  }
+
+  @Override
+  public String getOutFile() {
+    return this.outFile;
+  }
+
+  @Override
+  public void setOutFile(String outFile) {
+    this.outFile = outFile;
+  }
+
+  public static void main(String[] args) {
+    if (args.length != 3) {
+      throw new IllegalArgumentException("USAGE: JavaGrep regex rootPath outFile");
+    }
+
+    //Use default logger config
+    BasicConfigurator.configure();
+
+    JavaGrepImp javaGrepImp = new JavaGrepImp();
+    javaGrepImp.setRegex(args[0]);
+    javaGrepImp.setRootPath(args[1]);
+    javaGrepImp.setOutFile(args[2]);
+
+    try {
+      javaGrepImp.process();
+    } catch (Exception ex) {
+      javaGrepImp.logger.error(ex.getMessage(), ex);
```

Comment:
ex.getMessage() -> write a meaningful message instead as the second argument already captures the message